src/lib/storage/vector_compression/bitpacking/bitpacking_compressor.cpp

```diff
@@ -0,0 +1,39 @@
+#include "bitpacking_compressor.hpp"
+
+#include <algorithm>
+#include <cmath>
+
+#include "compact_vector.hpp"
+
+namespace opossum {
+
+class BitPackingVector;
+
+std::unique_ptr<const BaseCompressedVector> BitPackingCompressor::compress(const pmr_vector<uint32_t>& vector,
+                                                                           const PolymorphicAllocator<size_t>& alloc,
+                                                                           const UncompressedVectorInfo& meta_info) {
+  const auto max_element_it = std::max_element(vector.cbegin(), vector.cend());
+
+  auto required_bits = 1u;
+  if (max_element_it != vector.cend() && *max_element_it != 0) {
+    // add 1 to max_value because log2(1) = 0 but we need one bit to represent it.
+    required_bits = static_cast<uint32_t>(std::ceil(log2(*max_element_it + 1u)));
+  }
+
+  auto data = pmr_compact_vector<uint32_t>(required_bits, vector.size(), alloc);
+
+  // compactvector does not zero initialize it's memory, which leads to non-reproducible tests that
+  // use the binary writer (due to alignment, some memory at the end is not overwritten).
+  // Hence, fill the internal memory with zeroes.
+  std::fill_n(data.get(), data.bytes() / 8, 0);
```

Comment:
I have to revisit this part later. Not sure I understand the problems yet. Will recheck when I have seen all the code.

Reply:
let me know if i should try to explain the problem better

Reply:
Since there is no direct documentation: what does `.bytes()` return and why do you divide by 8?

Reply:
data.bytes() gives the number of allocated bytes for the internal memory. Added a comment that explains that. 
Regarding the 8: it was the word size of uint64_t in bytes. you're right, hardcoding it was bad style, changed it to `sizeof(*data.get())`

Reply:
Where does the `uint64_t` come from all of a sudden?

Reply:
Do you assume the written zero is a 64 bit value? If so, I'd make that clearer. Also, the way I read `std::fill_n` it takes an iterator,  not a pointer.

Besides that, if I follow that code correctly (good chances I am not, but that's a good way to check for sufficient documentation :D) you are dividing by `uint64_t` size and then write the resulting number of uint64_t elements. If the compact vector uses 65 bytes, you end up with 64 bytes written. What am I getting wrong here?

Reply:
the uint64_t is the word size that we template the compact_vector with in https://github.com/hyrise/hyrise/pull/2350/files#diff-88acbb51f256d7579751886ec22783b9ef4ed2ff16eaf886a05a8b5acc72358a . The compact_vector then also allocates its internal memory word-aligned, thus it can not have something like 65 bytes.

Reply:
I get that this is super ugly but unfortunately i did not find a better way to initialize this memory in a predictable way. If I don't initialize it, it still works functionally but a bunch of tests fail.

Reply:
Ok, the internal alignment is information that should be mentioned in a comment. Even though this raw pointer being used as in iterator should work, I am not really of fan of handling it like this.

Do you remember which tests failed?

Reply:
```
[  FAILED  ] 5 tests, listed below:
[  FAILED  ] BinaryEncodingTypes/BinaryWriterMultiEncodingTest.StringSegment/LZ4, where GetParam() = LZ4
[  FAILED  ] BinaryEncodingTypes/BinaryWriterMultiEncodingTest.AllTypesSegmentSorted/LZ4, where GetParam() = LZ4
[  FAILED  ] BinaryEncodingTypes/BinaryWriterMultiEncodingTest.AllTypesSegmentUnsorted/LZ4, where GetParam() = LZ4
[  FAILED  ] BinaryEncodingTypes/BinaryWriterMultiEncodingTest.AllTypesMixColumn/LZ4, where GetParam() = LZ4
[  FAILED  ] BinaryEncodingTypes/BinaryWriterMultiEncodingTest.AllTypesNullValues/LZ4, where GetParam() = LZ4
```
the errors boil down to: the bit-aligned values leave some small memory region at the end of the compact_vector's internal memory uninitialized and the binary_writer has to write out this unitialized memory because its output has to be compact_vector-word-aligned. I could try to fix this in the binary writer, but it would require more complicated bit-magic on this raw pointer there. Therefore, I thought it's best to solve this problem "at the source".

Reply:
Did you filter for LZ4? Curious why it only fails for LZ4. Do we also test Dictionary with CompactVector for binary writing?

Reply:
no I did not filter for LZ4. Yes, the same test also runs for all other encodings. I guess by chance the poslist fits nicely into the word size of the compact_vector so that no uninitialized memory is left in this specific test. Can you help me make a decision what the best solution for this is?

Reply:
We only test the binary import/export stuff for the default segment encodings (FSBA). Thus, only LZ4 tests can fail as it's the only one using compact vector.

Reply:
Ah, okay. I will then close this discussion as it doesn't seem like there is a better solution right now. Feel free to re-open if you have a better suggestion